### PR TITLE
Custom Decorators in Django

### LINKED ISSUE #7701: Creating and Utilizing Custom Decorators in Django
### Writing sample(s): 
- [How to Integrate Google Analytics in Typesense](https://aviyel.com/post/3169/how-to-integrate-google-analytics-in-typesense)
- [What Exactly is Multithreading](https://bovage.hashnode.dev/what-exactly-is-multithreading)
- [How to use PostgreSQL in a Django Project](https://bovage.hashnode.dev/how-to-use-postgresql-database-in-a-django-project)

#### Proposed title of the article
Creating and Utilizing Custom Decorators in Django

#### Proposed article introduction
In Python, a decorator is a function that takes another function as an argument and adds functionality or augments the function without changing it. Django, as a Python web framework, comes with a large number of built-in decorators. These built-in decorators are used for decorating function based views. However, the real-world application you're working on may require custom checks or validations that django doesn't supply out of the box.

In software development, situations like the one mentioned above are unavoidable. When such a situation happens while developing a Django application, the default option for the normal programmer is to conduct the checks or validations in the view. This method is straightforward, however it is inefficient, and it goes against software development best practices and principles such as DRY (Don't Repeat Yourself). This is because, as the program expands in size, the same check or validation must be performed in numerous views. As a result, the identical code is repeated across the application.

In this article, you'll learn how to develop custom decorators similar to the built-in decorators (such as login_required, require_http_methods, csrf_exempt) used in real-world applications. The following are custom decorators that will be created: 

- authentication_not_required: Users who are signed in will be unable to access a view as a result of this decorator. For login and registration views, this is useful..
- verification_required: Users who haven't verified their email address or phone number will be unable to access a view as a result of this.
- xhr_or_fetch_request_only:  This ensures only request via fetch, XHR(XMLHttpRequest) or AJAX(Asynchronously Javascript and XML) is allowed.

#### Key takeaways 
The reader should be able to do or understand the following after reading this article:
1. Have  a good understanding of how built-in django decorators work internally
2. How to write clean and readable codes that adhere to DRY(Don't Repeat Yourself) principle
3. How to use decorators in class based views

#### Article quality
The most effective approach to understand a programming topic is to apply it to real-world tasks.
This article describes and illustrates how the knowledge gained from this section article on [Getting Started with decorators in Python](https://www.section.io/engineering-education/python-decorators/) can be applied in real world applications. The codes written in this article are well-written and readable. It also delves into the inner workings of Django, utilizing core Python fundamentals.

#### References
[Django Docs](https://docs.djangoproject.com/en/4.0/)
[Django Codebase on Github](https://github.com/django/django)

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by BOVAGE**
## Our **Peer Reviewer and Content Moderator teams do NOT provide any revisions services.** All revisions and edits should be completed on your own forked repo (as to not take up room in the queue) in order for our team to review them all in a timely manner. 

**ANY ARTICLE SUBMITTED WITH GLARING ERRORS WILL BE IMMEDIATELY CLOSED.**

>As a rule of thumb - please be sure to only submit articles (pull requests) that are fully polished and ready to be published. Be sure to go through our resources documents for extra and 3rd party (vetted) resources to help improve overall technical writing.

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [ ] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

`This closes #7701`

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

**Comment by hectorkambow** [comment]
🚀

**Comment by hectorkambow** [comment]
@BOVAGE 

Question, does this pull request have a corresponding topic suggestion form?

**Comment by hectorkambow** [comment]
👍  

Thank you - just needed to make sure they were linked (see image) that way once the article is published (merged) it auto-closes the topic form as well. 

image

**Comment by BOVAGE** [comment]
> 👍
> 
> Thank you - just needed to make sure they were linked (see image) that way once the article is published (merged) it auto-closes the topic form as well.
> 
> image

Okay sir. Got it

**Comment by hectorkambow** [comment]
@BOVAGE just noticed that you have not included any author details (to create your author profile).

This is how your articles will be listed under your profile.

In order to publish the article we would need your author details - please see the new contributor folder for more info on how to submit.

**Comment by BOVAGE** [comment]
> @BOVAGE just noticed that you have not included any author details (to create your author profile).
> 
> This is how your articles will be listed under your profile.
> 
> In order to publish the article we would need your author details - please see the new contributor folder for more info on how to submit.

I added the author details now sir.

**Comment by WanjaMIKE** [comment]
Plagiarism check is done

**Comment by WanjaMIKE** [comment]
Hello @BOVAGE  I will get started on this technical review as soon as possible. Sorry for the delay. Thanks

**Comment by hectorkambow** [comment]
@Collince-Okeyo would like to do a test review on this PR 

@WanjaMIKE  if you have not started yet. 

👍

**Comment by WanjaMIKE** [comment]
Sure, @Collince-Okeyo you can take up the PR. Thanks

**Comment by hectorkambow** [comment]
🚀  

Thank you Mike 

@Collince-Okeyo  

@WanjaMIKE is a veteran in the program - he is always a reliable resource if you have any questions as well. 💯

**Comment by Collince-Okeyo** [comment]
> Sure, @Collince-Okeyo you can take up the PR. Thanks

Thank you @hectorkambow and @WanjaMIKE for this opportunity

File: content/articles/custom-decorators-in-django/index.md [hunk 1]
```diff
@@ -0,0 +1,499 @@
...
+```bash
+python manage.py createsuperuser
+python manage.py runserver
+```
+Go to your browser and navigate to http://127.0.0.1:8000/
+
+Check the image below
+![Authentication Decorators in Action](custom-decorators-in-django/authdec.gif)
+
+
```
Comment (Collince-Okeyo) [inline content/articles/custom-decorators-in-django/index.md:350]: Remove space

File: content/articles/custom-decorators-in-django/index.md [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+- A link or an OTP should be sent to the user's email address in order to verify his / her email account and activate the user. I have kept this tutorial simple that the admin should make the user active.
+> The default authentication backend does not return a user object for in active users. So if a user who created an account using the register.html won't be able to login unless the admin makes the user active.
```
Comment (Collince-Okeyo) [inline content/articles/custom-decorators-in-django/index.md:352]: Replace with...
The default authentication backend does not return a user object for inactive users. So a user who created an account using the 'register.html' won't be able to login unless the admin makes the user active.

File: content/articles/custom-decorators-in-django/index.md [hunk 3]
```diff
@@ -0,0 +1,499 @@
...
+
+Now to implement the xhr_request_only decorator created earlier.
+Inside posts/views.py
```
Comment (Collince-Okeyo) [inline content/articles/custom-decorators-in-django/index.md:355]: Replace with...
Now to implement the 'xhr_request_only' decorator created earlier, add the following code inside 'posts/views.py.'

File: content/articles/custom-decorators-in-django/index.md [hunk 4]
```diff
@@ -0,0 +1,499 @@
...
+    elif action == "downvote":
+        post.upvote_users.remove(request.user)
+        post.downvote_users.add(request.user)
+    return JsonResponse({"message": action})
+```
+In the above code,
+
+- Line 3 is to import xhr_request_only from the decorators module.
+- Line 4 is to import verification_required from the decorators module in the accounts application.
+- The built-in require_POST decorator has been used on the post_vote_view to ensure  that it only accepts the POST request method.
```
Comment (Collince-Okeyo) [inline content/articles/custom-decorators-in-django/index.md:391]: Replace with...
The built-in require_POST decorator has been used on the 'post_vote_view' to ensure that it only accepts the 'POST' request method.

**Review by Collince-Okeyo — CHANGES_REQUESTED** [review]
@BOVAGE ,
Please correct the suggested corrections then let me know to proceed with the review. Thank you...

**Comment by BOVAGE** [comment]
@Collince-Okeyo, I have made the suggested changes. Thank you.

**Review by Collince-Okeyo — APPROVED** [review]

**Comment by hectorkambow** [comment]
@Collince-Okeyo 

It looks like this PR is ready for a "ready for final review" label? 

Assuming both prelim & tech review have been completed?

**Comment by Collince-Okeyo** [comment]
> @Collince-Okeyo
> 
> It looks like this PR is ready for a "ready for final review" label?
> 
> Assuming both prelim & tech review have been completed?

@hectorkambow , thank you for following up,...

The PR is ready for a "ready for final review" label since all the reviews are completed.

**Comment by Collince-Okeyo** [comment]
Hello @hectorkambow ,

Please add the 'ready for final review' tag to this PR

**Comment by BOVAGE** [comment]
Hello @hectorkambow.

@Collince-Okeyo is done with the review.

**Comment by hectorkambow** [comment]
Thank you for the ping @BOVAGE  

I lose many notification in my inbox.

**Comment by hectorkambow** [comment]
@BOVAGE 
Well done ! 

Please review any edits made by your peer reviewer (or myself) to improve future blog post.

You'll be able to see your live published article on the site (in about 30 mins or so).

To give your article a boost in page engagement - be sure to share it across your social site (LinkedIn/Twitter) and tag the EngEd program so we can also share it and give it some traction/exposure as well!

If you have not done so already - consider adding a skill section to your author page. This will go a long way for potential future employers.

NOTE: Also, take into consideration that you can add any other past projects of yours: videos, talks, tutorials, articles (on ANY other site) - you can add them to your EngEd profile.

Think of the EngEd profile as your one-stop shop for everything you do in DEV - making it easier for you to share all your accomplishments when you go out for an interview or employment opp.

It can be from any blog site (or site in general).

Let me know if you have any questions moving forward.

Also important NOTE: We wanted to mention that we're going to take a focus on content centered around Kubernetes (K8s) for the month of August.

The best five (5) K8s related articles that come in this month will be eligible to receive a complimentary virtual pass to [KubeCon USA](https://events.linuxfoundation.org/kubecon-cloudnativecon-north-america/) ($75.00 value) in Oct.

We're looking for content that enriches developers' understanding of K8s, best practices, and its many use cases.
Please be sure to go over the already [published content the EngEd community has created on Kubernetes](https://www.section.io/engineering-education/search/?q=kubernetes) to see where we can build upon or further explore. 🚀

**Pull request merged**